src/pages/problems.tsx

```diff
@@ -1,47 +1,126 @@
-const CustomStatusSelection = connectRefinementList(Status);
-const CustomSourceSelection = connectRefinementList(Source);
-const CustomSearchBox = connectSearchBox(SearchBox);
-const CustomHits = connectHits(ProblemHits);
-const CustomRefinementList = connectRefinementList(RefinementList);
+type DataProps = {
+  allProblemInfo: {
+    nodes: {
+      uniqueId: string;
+    }[];
+  };
+};
 
-export default function ProblemsPage(props: PageProps) {
-  const { problems } = props.data as any;
-  const problemIds = problems.edges.reduce((acc, cur) => {
-    const problem = cur.node;
-    acc.push(problem.uniqueId);
-    return acc;
-  }, []);
+export default function ProblemsPage(props: PageProps<DataProps>) {
+  const {
+    allProblemInfo: { nodes: problems },
+  } = props.data;
+  const problemIds = problems.map(problem => problem.uniqueId);
+  const userProgress = useUserProgressOnProblems();
+  const selectionMetadata: SelectionProps[] = [
+    {
+      attribute: 'difficulty',
+      limit: 500,
+      placeholder: 'Difficulty',
+      searchable: false,
+      isMulti: true,
+    },
+    {
+      attribute: 'problemModules.title',
+      limit: 500,
+      placeholder: 'Modules',
+      searchable: true,
+      isMulti: true,
+    },
+    {
+      attribute: 'source',
+      limit: 500,
+      placeholder: 'Source',
+      searchable: true,
+      isMulti: true,
+    },
+    {
+      attribute: 'isStarred',
+      limit: 500,
+      placeholder: 'Starred',
+      searchable: false,
+      transformLabel: label => (label == 'true' ? 'Yes' : 'No'),
+      isMulti: false,
+    },
+    {
+      attribute: 'problemModules.id',
+      limit: 500,
+      placeholder: 'Section',
+      searchable: false,
+      isMulti: true,
+      items: [
+        {
+          label: 'General',
+          value: SECTIONS.general.map(chapter => chapter.items).flat(),
+        },
+        {
+          label: 'Bronze',
+          value: SECTIONS.bronze.map(chapter => chapter.items).flat(),
+        },
+        {
+          label: 'Silver',
+          value: SECTIONS.silver.map(chapter => chapter.items).flat(),
+        },
+        {
+          label: 'Gold',
+          value: SECTIONS.gold.map(chapter => chapter.items).flat(),
+        },
+        {
+          label: 'Platinum',
+          value: SECTIONS.plat.map(chapter => chapter.items).flat(),
+        },
+        {
+          label: 'Advanced',
+          value: SECTIONS.adv.map(chapter => chapter.items).flat(),
+        },
+      ],
+    },
+    {
+      attribute: 'objectID',
+      limit: 500,
+      placeholder: 'Status',
+      searchable: false,
+      isMulti: true,
+      items: [
+        'Not Attempted',
+        'Solving',
+        'Reviewing',
+        'Skipped',
+        'Ignored',
+        'Solved',
+      ].map(label => ({
+        label,
+        value: problemIds.filter(
+          id => (userProgress[id] ?? 'Not Attempted') == label
```

Comment:
`Not Attempted` seems like it might be an expensive query (ie. the list will be very long). Does it make more sense to *exclude* ID's specifically for `Not Attempted`?

Reply:
Hm, I don't think we can use exclusion in conjunction with inclusion but my refactor to use `setIndexUiState` vs multiple `refine` calls should mitigate this issue

Reply:
> I don't think we can use exclusion in conjunction with inclusion

Why not? Does adding a `-` in front of every refinement not work (https://discourse.algolia.com/t/facet-filter-negative-and-not-in-react-instant-search/4470)? (I haven't actually tried it so it's quite possible it doesn't work...)

> should mitigate this issue

Is there a limit of 1000 refinements for Algolia? If so, would this break if we exceed 1000 problems in the USACO Guide? (If we aren't that close to hitting the limit, it's probably fine as-is though.)

If we are close to hitting the limit, I think the "proper" solution will be to not allow searching by name in addition to searching by other properties. If we only search by name, we can use Algolia; if we only search by other properties like module, difficulty, and solve status, we can just use plain Javascript to make a list of all problem IDs that meet the criterion.

I can't think of a use case where a user will search by both name and other properties. The most likely use case I'm imagining is searching for a specific problem (searching by name) or searching to create a problemset (searching by properties like module, difficulty, solve status).

(We don't need to build this out now though)

Reply:
> Why not? Does adding a - in front of every refinement not work (https://discourse.algolia.com/t/facet-filter-negative-and-not-in-react-instant-search/4470)? (I haven't actually tried it so it's quite possible it doesn't work...)

I think this filters syntax ("AND", "NOT", "-", etc.) is unfortunately not generally available in Algolia's React/Hooks API (`useSearchBox`, `useRefinementList`, etc).

> If we are close to hitting the limit, I think the "proper" solution will be to not allow searching by name in addition to searching by other properties. If we only search by name, we can use Algolia; if we only search by other properties like module, difficulty, and solve status, we can just use plain Javascript to make a list of all problem IDs that meet the criterion.

Hm, we could probably build the filters directly into `ProblemHits` if necessary, although that does get rid of some small Algolia features like filter options changing dynamically based on the other filter states (in either case, we can definitely allow both searching by name and by property 😄 )